### PR TITLE
Implement merge sets operation for Build IDs

### PR DESCRIPTION
**Pull request by Sushisource**
**What changed?**
Implement the operation added in https://github.com/temporalio/api/pull/297

**Why?**
Gotta implement bro

**How did you test it?**
Added UTs

**Potential risks**

**Is hotfix candidate?**

### CODE REVIEW

File: service/matching/version_sets_test.go [hunk 1]
```diff
@@ -644,7 +680,68 @@ func TestGetBuildIdDeltas(t *testing.T) {
...
+	req := mkMergeSet("1", "2")
+	nextClock := hlc.Next(clock, commonclock.NewRealTimeSource())
+	updatedData, err := UpdateVersionSets(nextClock, initialData, req, 0, 0)
+	assert.NoError(t, err)
+	// Should only be three sets now
+	assert.Equal(t, 3, len(updatedData.VersionSets))
+	// The overall default set should not have changed
+	assert.Equal(t, "3", updatedData.GetVersionSets()[2].GetBuildIds()[0].Id)
+	// But set 1 should now have 2, maintaining 1 as the default ID
+	assert.Equal(t, "1", updatedData.GetVersionSets()[1].GetBuildIds()[1].Id)
```
Comment (dnr) [inline service/matching/version_sets_test.go:703]: can you verify that it has two set ids and they match the individual set ids from before?

File: service/matching/version_sets.go [hunk 2]
```diff
@@ -265,13 +289,18 @@ func extractTargetedVersion(req *workflowservice.UpdateWorkerBuildIdCompatibilit
 		return req.GetPromoteSetByBuildId()
 	} else if req.GetPromoteBuildIdWithinSet() != "" {
 		return req.GetPromoteBuildIdWithinSet()
+	} else if req.GetAddNewBuildIdInNewDefaultSet() != "" {
+		return req.GetAddNewBuildIdInNewDefaultSet()
 	}
-	return req.GetAddNewBuildIdInNewDefaultSet()
+	return req.GetMergeSets().GetPrimarySetBuildId()
```
Comment (dnr) [inline service/matching/version_sets.go:295]: I'd still rather extract all the args within the big switch but :shrug:

File: service/matching/version_sets_test.go (hunk 1 above)
Comment (Sushisource) [reply service/matching/version_sets_test.go:703]: Added

**Review by dnr — APPROVED** [review]

File: service/matching/version_sets.go [hunk 3]
```diff
@@ -253,6 +253,30 @@ func updateImpl(timestamp hlc.Clock, existingData *persistencespb.VersioningData
...
+			return existingData, nil
+		}
+		// Merge the sets together, preserving the primary set's default by making it have to most recent timestamp.
+		primarySet := modifiedData.VersionSets[targetSetIdx]
+		primarySet.DefaultUpdateTimestamp = &timestamp
+		justPrimaryData := &persistencespb.VersioningData{
+			VersionSets:            []*persistencespb.CompatibleVersionSet{primarySet},
+			DefaultUpdateTimestamp: modifiedData.DefaultUpdateTimestamp,
+		}
+		secondarySet := modifiedData.VersionSets[secondarySetIdx]
```
Comment (bergundy) [inline service/matching/version_sets.go:276]: Same don't mutate the original data

File: service/matching/version_sets.go [hunk 4]
```diff
@@ -253,6 +253,30 @@ func updateImpl(timestamp hlc.Clock, existingData *persistencespb.VersioningData
...
+		if secondarySetIdx == -1 {
+			return nil, serviceerror.NewNotFound(fmt.Sprintf("targeted secondary version %v not found", secondaryBuildID))
+		}
+		if targetSetIdx == secondarySetIdx {
+			// Nothing to be done
+			return existingData, nil
+		}
+		// Merge the sets together, preserving the primary set's default by making it have to most recent timestamp.
+		primarySet := modifiedData.VersionSets[targetSetIdx]
+		primarySet.DefaultUpdateTimestamp = &timestamp
```
Comment (bergundy) [inline service/matching/version_sets.go:271]: Don't mutate the original, copy the set first.

File: service/matching/version_sets.go [hunk 5]
```diff
@@ -265,13 +289,18 @@ func extractTargetedVersion(req *workflowservice.UpdateWorkerBuildIdCompatibilit
...
+		return req.GetAddNewBuildIdInNewDefaultSet()
 	}
-	return req.GetAddNewBuildIdInNewDefaultSet()
+	return req.GetMergeSets().GetPrimarySetBuildId()
 }
 
 // Finds the version in the version sets, returning (set index, index within that set)
 // Returns -1, -1 if not found.
 func findVersion(data *persistencespb.VersioningData, buildID string) (setIndex, indexInSet int) {
+	if buildID == "" {
```
Comment (bergundy) [inline service/matching/version_sets.go:301]: When does this happen?

File: service/matching/version_sets_test.go [hunk 6]
```diff
@@ -644,7 +680,71 @@ func TestGetBuildIdDeltas(t *testing.T) {
 }
 
 func TestGetBuildIdDeltas_AcceptsNils(t *testing.T) {
+	t.Parallel()
 	added, removed := GetBuildIdDeltas(nil, nil)
 	assert.Equal(t, []string(nil), removed)
 	assert.Equal(t, []string(nil), added)
 }
+
+func TestMergeSets(t *testing.T) {
```
Comment (bergundy) [inline service/matching/version_sets_test.go:689]: Please test that the original data doesn't get mutated.

File: service/matching/version_sets.go (hunk 4 above)
Comment (Sushisource) [reply service/matching/version_sets.go:271]: This is the "modified" data, which is already a copy. Why would we need to copy it again?

File: service/matching/version_sets.go (hunk 5 above)
Comment (Sushisource) [reply service/matching/version_sets.go:301]: If the "secondary" set id isn't specified by the user. It's just a reasonable check in any case.

File: service/matching/version_sets.go (hunk 4 above)
Comment (Sushisource) [reply service/matching/version_sets.go:271]: Ah ok nevermind I see the initial copy is still keeping pointers to all the internal sets

File: service/matching/version_sets.go (hunk 5 above)
Comment (bergundy) [reply service/matching/version_sets.go:301]: WDYM by "secondary" set ID? this checks `buildID`.

Comment (Sushisource) [reply service/matching/version_sets.go:301]: That is the path that would be taken to end up calling this with an empty build id.

**Pull request merged**